contrib/ros/envs/sawyer_env.py

```diff
@@ -0,0 +1,356 @@
+from rllab.misc import logger
+from rllab.spaces import Box
+
+from contrib.ros.envs.ros_env import RosEnv
+from contrib.ros.robots.sawyer import Sawyer
+from contrib.ros.util.common import rate_limited
+from contrib.ros.util.gazebo import Gazebo
+
+import intera_interface
+
+
+def goal_distance(goal_a, goal_b):
+    """
+    :param goal_a:
+    :param goal_b:
+    :return: distance between goal_a and goal_b
+    """
+    assert goal_a.shape == goal_b.shape
+    return np.linalg.norm(goal_a - goal_b, axis=-1)
+
+
+class SawyerEnv(RosEnv, Serializable):
+    """
+    ROS Sawyer Env
+    """
+
+    def __init__(self,
+                 initial_robot_joint_pos,
+                 robot_control_mode,
+                 real,
+                 has_object,
+                 distance_threshold,
+                 initial_model_pos,
+                 sparse_reward,
+                 target_in_the_air,
+                 target_range=0.15,
+                 obj_range=0.15):
+        """
+        :param initial_robot_joint_pos: {str: float}
+                {'joint_name': value}
+        :param robot control mode: str
+                'effort'/'position'/'velocity'
+        :param real: bool
+                if the environment is for real robot or simulation
+        :param has_object: bool
+                if there is object in this experiment
+        :param distance_threshold: float
+                When the distance from robot endpoint to the target point is smaller than this threshold
+                the training episode is done.
+                To calculate the sparse reward, if the distance is smaller than this threshold returns -1
+                else returns 0.
+        :param initial_model_pos: dict
+                joint positions and object positions
+        :param sparse_reward: bool
+                if reward_type is 'sparse'
+        :param target_in_the_air: bool
+                set to see if the target can be in the air
+        :param target_range: float
+                the new target position for the new training episode would be randomly sampled inside
+                [initial_gripper_pos - target_range, initial_gripper_pos + target_range]
+        :param obj_range: float
+                the new initial object x,y position for the new training episode would be randomly sampled inside
+                [initial_gripper_pos - obj_range, initial_gripper_pos + obj_range]
+        """
+        # run the superclass initialize function first
+        Serializable.quick_init(self, locals())
+
+        logger.log('Initializing ros node!')
+        rospy.init_node('SawyerEnv', anonymous=True)
+        # Verify robot is enabled
+        self.limb = intera_interface.Limb('right')
+        logger.log('Initializing robot!')
+        self._robot = Sawyer(
+            initial_robot_joint_pos, control_mode=robot_control_mode)
+        if not self._robot.enabled:
+            logger.log("Please enable the sawyer!")
+            raise Exception
```

Comment:
1. Don't raise anonymous exceptions--it should have some string description of what went wrong.
1. I think you should use `RuntimeError` for this.
1. Please move the long-running ROS/robot setup code to a function `initialize()` and make it part of the `RosEnv` interface.